spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -68,18 +65,31 @@
 
 	private final SubscriberFactory subscriberFactory;
 
+	private final SubscriberStub subscriberStub;
+
+	private PubSubAcknowledger acknowledger;
+
+	public PubSubTemplate(PublisherFactory publisherFactory, SubscriberFactory subscriberFactory) {
+		this(publisherFactory, subscriberFactory,
+				new DefaultPubSubAcknowledger(subscriberFactory.createSubscriberStub()));
+	}
+
 	/**
-	 * {@link PubSubTemplate} constructor that uses a {@link JacksonPubSubMessageConverter} to
-	 * serialize and deserialize payloads.
+	 * Default {@link PubSubTemplate} constructor that uses a {@link JacksonPubSubMessageConverter}
+	 * to serialize and deserialize payloads.
 	 * @param publisherFactory the {@link com.google.cloud.pubsub.v1.Publisher} factory to
 	 * publish to topics
 	 * @param subscriberFactory the {@link com.google.cloud.pubsub.v1.Subscriber} factory
 	 * to subscribe to subscriptions
+	 * @param acknowledger auto-acknowledges pulled messages
 	 */
 	public PubSubTemplate(PublisherFactory publisherFactory,
-			SubscriberFactory subscriberFactory) {
+			SubscriberFactory subscriberFactory,
+			PubSubAcknowledger acknowledger) {
```

Comment:
As I said I'm not sure in this extra argument since there is no guarantee that `PubSubAcknowledger` is going to be based on the provided `SubscriberFactory` as well.

However I think we need an API to expose this internal `PubSubAcknowledger` instance to let end-user to really perform a batch (n)ack. Not sure why the Pub/Sub contract is to force end-user to group by `subscription` and not `multi` for ordered poll, but let it be. My point is that we need somehow to expose hook to let end-users to do that and have not only per message basis.

Reply:
The API is exposed, `GcpPubSubAutoConfiguration` provides a `PubSubAcknowledger` that can be used for bulk acking.
That's how I intended users to get the `PubSubAcknowledger` from.

Reply:
Yeah... That's confusing and not so flexible.
I suggest just expose `getAkcnowlader()` on the `SubscriberFactory`: you can't be sure that user doesn't have several different `SubscriberFactory`. Therefore no beans for the `Akcnowlader`, please.